### PR TITLE
Update Mle and Joiner implementations

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two different (related) commits:

**[joiner] update Joiner implementations**

This commit contains the following changes to `Joiner` class
implementations:

- Notifier events are signaled on all `Joiner` state changes
  and are logged (through use of `Joiner::SetState()`).
- Discover scan during join uses filtering (removes the need
  to perform filtering by the `Joiner` class itself).
- `Joiner::Stop()` is changed to allow it to be called at any
  stage of join process and it correctly cleans all states.
- Priority assignment to discovered networks is updated/fixed
  (prioritizing networks with an exact match of Joiner ID in their
  Steering Data compared to ones that allow all joiners, and
  sub-prioritize based on signal strength).
- Logs are updated.

-------

**[mle] update discover scan filtering**

This commit makes improvements to MLE Discover Scan implementation,
mainly the steering data filtering logic:
- The SHA-256 of EUI64 (Joiner ID) and then the related bloom filter
  CRC-CCIT and CRC-ANSI indexes are calculated and saved at start of
  scan operation (removing the need to re-calculate them on every scan
  result).
- When filtering is enabled, a Discover Scan result with no Steering
  Data is now rejected.
- `Mle` class member variables related to discover scan operation
  are renamed to start with `mDisocver` prefix.

### CODE REVIEW

File: src/core/meshcop/joiner.cpp [hunk 1]
```diff
@@ -95,92 +103,125 @@ otError Joiner::Start(const char *     aPSKd,
     ThreadNetif &   netif = GetNetif();
     otError         error;
     Mac::ExtAddress joinerId;
-    Crc16           ccitt(Crc16::kCcitt);
-    Crc16           ansi(Crc16::kAnsi);
 
-    VerifyOrExit(mState == OT_JOINER_STATE_IDLE, error = OT_ERROR_BUSY);
+    otLogInfoMeshCoP("Joiner starting");
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: nit: move below `VerifyOrExit()`?

File: src/core/meshcop/joiner.cpp [hunk 2]
```diff
@@ -95,92 +103,125 @@ otError Joiner::Start(const char *     aPSKd,
...
+    if (aRssi == OT_RADIO_RSSI_INVALID)
     {
-        error = TryNextJoin();
+        aRssi = -127;
     }
 
-    if (error == OT_ERROR_NOT_FOUND && mCallback)
+    // Limit the RSSI to range (-128, 0), i.e. -128 < aRssi < 0.
+
+    if (aRssi == -128)
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: nit: use `<=` in case type changes?

File: src/core/meshcop/joiner.cpp [hunk 3]
```diff
@@ -190,118 +231,130 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
...
 
-        if (error != OT_ERROR_NONE)
+    otLogInfoMeshCoP("Joiner discover network: %s, pan:0x%04x, port:%d, chan:%d, rssi:%d, allow-any:%s",
+                     static_cast<const Mac::ExtAddress &>(aResult.mExtAddress).ToString().AsCString(), aResult.mPanId,
+                     aResult.mJoinerUdpPort, aResult.mChannel, aResult.mRssi, doesAllowAny ? "yes" : "no");
+
+    priority = CalculatePriority(aResult.mRssi, doesAllowAny);
+
+    // Find the lowest priority entry in the list
+    for (JoinerRouter *cur = &mJoinerRouters[0]; cur < &mJoinerRouters[OT_ARRAY_LENGTH(mJoinerRouters)]; cur++)
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: What about keeping this list sorted based on priority? Should simplify `GetNextJoinerRouter()`.

File: src/core/meshcop/joiner.cpp [hunk 4]
```diff
@@ -314,25 +367,29 @@ void Joiner::HandleSecureCoapClientConnect(bool aConnected, void *aContext)
...
+        mTimer.Start(kReponseTimeout);
+    }
+    else
+    {
+        JoinerRouter *router = GetNextJoinerRouter();
+
+        while (router != NULL)
         {
-            Complete(OT_ERROR_SECURITY);
+            VerifyOrExit(Connect(*router) != OT_ERROR_NONE);
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: Separate this while-loop into a separate method to share with similar code above?

Comment (abtink) [reply src/core/meshcop/joiner.cpp]: :+1: Changed this in the new pushed commit. Added some new code to handle the previous error (from last attempt).

File: src/core/meshcop/joiner.cpp (hunk 2 above)
Comment (abtink) [reply src/core/meshcop/joiner.cpp]: Changed in latest push.

File: src/core/meshcop/joiner.cpp [hunk 5]
```diff
@@ -95,92 +103,125 @@ otError Joiner::Start(const char *     aPSKd,
...
+    // Use extended address based on factory-assigned IEEE EUI-64
     GetJoinerId(joinerId);
     netif.GetMac().SetExtAddress(joinerId);
     netif.GetMle().UpdateLinkLocalAddress();
 
-    for (size_t i = 0; i < sizeof(joinerId); i++)
+    SuccessOrExit(error = netif.GetCoapSecure().Start(kJoinerUdpPort));
+    SuccessOrExit(error = netif.GetCoapSecure().SetPsk(reinterpret_cast<const uint8_t *>(aPSKd),
+                                                       static_cast<uint8_t>(strlen(aPSKd))));
+    SuccessOrExit(error = netif.GetCoapSecure().GetDtls().mProvisioningUrl.SetProvisioningUrl(aProvisioningUrl));
```
Comment (bukepo) [inline src/core/meshcop/joiner.cpp]: I think we should submit another PR to move `SetProvisioningUrl()` out of `Dtls`.

File: src/core/meshcop/joiner.cpp [hunk 6]
```diff
@@ -583,8 +641,52 @@ void Joiner::HandleTimer(void)
         break;
     }
 
-    Complete(error);
+    Finish(error);
+}
+
+const char *Joiner::JoinerStateToString(otJoinerState aState)
```
Comment (bukepo) [inline src/core/meshcop/joiner.cpp]: I am wondering if we should use template to do this kind of job.
```cpp
template <int kState> const char *Joiner::JoinerStateToString(void);
template <> const char *Joiner::JoinerStateToString< OT_JOINER_STATE_IDLE >(void) { return "Idle"; }
```
This could save code if the string is not used. thoughts @abtink ?

File: src/core/thread/mle.cpp [hunk 7]
```diff
@@ -506,9 +508,29 @@ otError Mle::Discover(const Mac::ChannelMask &aScanChannels,
     MeshCoP::DiscoveryRequestTlv discoveryRequest;
     uint16_t                     startOffset;
 
-    VerifyOrExit(!mIsDiscoverInProgress, error = OT_ERROR_BUSY);
+    VerifyOrExit(!mDiscoverInProgress, error = OT_ERROR_BUSY);
 
-    mEnableEui64Filtering = aEnableEui64Filtering;
+    mDiscoverEnableFiltering = aEnableFiltering;
```
Comment (bukepo) [inline src/core/thread/mle.cpp]: Is it better to implement this filtering in joiner(`Joiner::HandleDiscoverResult()`)? I think this keeps the interface clean and is more extensible.

File: src/core/meshcop/joiner.cpp (hunk 3 above)
Comment (abtink) [reply src/core/meshcop/joiner.cpp]: Not sure if it makes the code much simpler. 
I added the logic to keep it sorted in a new commit here: https://github.com/openthread/openthread/pull/3721/commits/d93799f7f87696a41fae45fe5762e7d7e6783964

@jwhui, @bukepo  what do you think?

File: src/core/meshcop/joiner.cpp (hunk 6 above)
Comment (abtink) [reply src/core/meshcop/joiner.cpp]: Not sure if I follow. How do we use this model?
I think the template argument need to be a compile-time constant. We cannot call it with a variable like `JoineStateToString<aState>`?

File: src/core/thread/mle.cpp (hunk 7 above)
Comment (abtink) [reply src/core/thread/mle.cpp]: The EUI64 filtering was already added to MLE discover scan from https://github.com/openthread/openthread/pull/1685 along with OOB steering data control.  I am just updating the existing implementation in this PR (first commit).

Note that the MLE discover scan EUI64 filtering is already part of the public API `otThreadDiscover(.., aEnableEui64Filtering,...)` and exposed through spinel and wpantund. So we don't want to  remove it from MLE discover scan (IIRC, this feature was added (for internal use) to be able to filter scan result during weave pairing).

In this PR, we are actually removing the redundant filtering code in `Joiner` and use the existing one in MLE discover scan.

File: src/core/thread/mle.cpp [hunk 8]
```diff
@@ -3730,33 +3753,13 @@ otError Mle::HandleDiscoveryResponse(const Message &aMessage, const Ip6::Message
...
-                    ansi.Update(extaddr.m8[i]);
-                }
-
-                // Drop responses that don't match the bloom filter
-                if (!steeringData.GetBit(ccitt.Get() % steeringData.GetNumBits()) ||
-                    !steeringData.GetBit(ansi.Get() % steeringData.GetNumBits()))
-                {
-                    ExitNow(error = OT_ERROR_NONE);
-                }
+                VerifyOrExit((steeringData.GetBit(mDiscoverCcittIndex % steeringData.GetNumBits()) &&
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/thread/mle.cpp]: We don't want to return an error here since we log the error at `exit` label (at warning level). The Discovery response itself is not bad or incorrect. We just filtered it (we are not looking for this network). 

We can log the fact that steering data did include this device and we are filtering the scan result, but not sure how useful it would be (we anyways log the results that pass filer in `Joiner` itself). 
IIRC, we used to log a similar log line when we filtered non-joinable networks and it misled a lot folks looking at the logs during pairing thinking there was an error with scan operation (specially in test areas where they are many networks and scan result can be large).

**Comment by abtink** [comment]
@librasungirl, you are correct, however, the new design of not trying next joiner routers after we pass  `OT_JOINER_STATE_CONNECTED` state (i.e. we successfully connect to one over DTLS) is intentional. Basically if we can successfully establish a session through a `JoinerRouter` with commissioner but fail at the next steps (say the `JoinerFinalize` response) we would consider the whole Join attempt a failure. This seems to align with the spec, but there are also other reasons why we went for this. 

We want to remove the requirement for the vendor related info to be `static` pointers (without the need to have buffers/memory space to store them). Note that these parameters are only used in "Joiner Finalize req" message. The idea is to prepare the the "Joiner Finalize" message from `Start()` call, keep track of message pointer, and once we get to `CONNECTED` state we send the message (which basically consume and free the message). With this model, we cannot try out another Joiner Router since the message is already consumed. I plan to submit follow-up PRs related to this after this.

Thoughts? Concerns?

File: src/core/meshcop/joiner.cpp [hunk 9]
```diff
@@ -433,20 +503,18 @@ void Joiner::HandleJoinerFinalizeResponse(Coap::Message &         aMessage,
...
-    aMessage.Read(aMessage.GetOffset(), aMessage.GetLength() - aMessage.GetOffset(), buf);
-    otDumpCertMeshCoP("[THCI] direction=recv | type=JOIN_FIN.rsp |", buf, aMessage.GetLength() - aMessage.GetOffset());
+#if OPENTHREAD_ENABLE_CERT_LOG
+    LogCertMessage("[THCI] direction=recv | type=JOIN_FIN.rsp |", aMessage);
 #endif
 
 exit:
-    Close();
+    GetNetif().GetCoapSecure().Disconnect();
+    GetNetif().GetIp6Filter().RemoveUnsecurePort(kJoinerUdpPort);
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/meshcop/joiner.cpp]: We would then keep session alive for longer. Is there any advantage for letting timer clean this? With this code, it's explicit that we are done and can close the session.

File: src/core/meshcop/joiner.cpp [hunk 10]
```diff
@@ -190,118 +232,154 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
...
+        otError       error;
+
+        if (router.mPriority == 0)
         {
-            joinerRouter = &mJoinerRouters[i];
+            break;
         }
 
-        if (joinerRouter->mPriority == 0)
+        error = Connect(router);
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/meshcop/joiner.cpp]: Please check comment below.

File: src/core/meshcop/joiner.cpp [hunk 11]
```diff
@@ -187,117 +228,154 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
...
-    {
-        otError error = TryNextJoin();
 
-        if (error != OT_ERROR_NONE)
+    VerifyOrExit(entry < end);
+
+    // Find the last in-use element in the list after entry
+    // (in-use elements have non-zero priority).
+
+    for (last = entry; last < end; last++)
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: If the goal is to minimize code, you could avoid searching for last and move the entire remainder of the list (which currently defaults to 2 entries).

Comment (abtink) [reply src/core/meshcop/joiner.cpp]: I agree... should be ok and would save us a bunch of lines.
Changed this in new pushed commit (also squashed the logic to keep the list sorted into the main commit).

**Comment by librasungirl** [comment]
@abtink , if only speaking of the spec, the joiner hunting algorithm in 8.4.4.1.2 seems in favor of keep trying next available joiner router until success.
> b. If the DTLS handshake fails or there is a timeout, move on to the next network.

For the requirement you described, maybe clone the FIN.req message for one Joiner Router try - which would consume more message buffer space instead. Sorry that I don't have any better idea.

**Comment by abtink** [comment]
Thanks @librasungirl

With new model, we would still try the next joiner if DTLS connection itself fails/timeout (from the `HandleSecureCoapClientConnect()` with `aConnected` being false). Note that we don't start the `mTimer` in `Joiner` class till after that. So we rely on `Dtls` class for timeout of DTLS connection.

Only if DTLS is successfully established (we reached our intended commissioner and established a secure connection to it) and then something fails afterwards we don't try other routers/networks.  In spec, section 8.4.5 we seem to have this statement for when JOIN_FIN.rsp contains a reject status (or I assume we timeout and never get JOIN_FIN.res): 
> The Joiner reattempts joining the same network anticipating that the Commissioner is now running the correct application. 

If I understand this correctly, we need to try to connect to same router (same network) again so I guess we don't need to try other routers (other networks) after this?

I think this is all rather special case (considering we currently have 2 joiner-router entries)... :)
